spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/ApiKeyFetcher.java

```diff
@@ -0,0 +1,102 @@
+package eu.coatrack.proxy.security;
+
+/*-
+ * #%L
+ * coatrack-proxy
+ * %%
+ * Copyright (C) 2013 - 2021 Corizon | Institut für angewandte Systemtechnik Bremen GmbH (ATB)
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+import eu.coatrack.api.ApiKey;
+import eu.coatrack.proxy.security.exceptions.ApiKeyFetchingFailedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Offers request services to the Coatrack admin server to receive single
+ * API keys or a list of API keys for all services offered by this gateway.
+ *
+ * @author Christoph Baier
+ */
+
+@Service
+public class ApiKeyFetcher {
+
+    private static final Logger log = LoggerFactory.getLogger(ApiKeyFetcher.class);
+
+    private final UrlResourcesProvider urlResourcesProvider;
+    private final RestTemplate restTemplate;
+
+    public ApiKeyFetcher(RestTemplate restTemplate, UrlResourcesProvider urlResourcesProvider) {
+        this.restTemplate = restTemplate;
+        this.urlResourcesProvider = urlResourcesProvider;
+    }
+
+    public List<ApiKey> requestLatestApiKeyListFromAdmin() throws ApiKeyFetchingFailedException {
+        log.debug("Requesting latest API key list from CoatRack admin.");
+
+        ResponseEntity<ApiKey[]> responseEntity;
+        try {
+            responseEntity = restTemplate.getForEntity(
+                    urlResourcesProvider.getApiKeyListRequestUrl(), ApiKey[].class);
+        } catch (RestClientException e) {
+            throw new ApiKeyFetchingFailedException("Trying to request the latest API key list from Admin, the " +
+                    "connection failed." + e.getMessage() + e);
+        }
+
+        ApiKey[] apiKeys = (ApiKey[]) extractBodyFromResponseEntity(responseEntity);
```

Comment:
The casting can not fail because the RestTemplate assures type correctness using the ApiKey[] argument:
```restTemplate.getForEntity(urlResourcesProvider.getApiKeyListRequestUrl(), ApiKey[].class);```

Reply:
The purpose of the ```extractBodyFromResponseEntity``` is to provide both, an API key and an API key list which avoids code duplication. That is why casting should take place outside of that method.

Reply:
OK, then the casting should be moved into the try block, as it depends on the RestTemplate call. As far as I can see, this was already done, so :heavy_check_mark: